### PR TITLE
Remove goreleaser

### PR DESCRIPTION
**Pull request by danielrbradley**
The first issue we have is that goreleaser creates a bottleneck building each binary in series on the same machine. We can build each platform on a different job because goreleaser won’t let us bring existing binaries without using the paid product. This affects other repositories more than azure-native, but this work paves the way for creating an improved pattern to adopt elsewhere.

The second issue with this requirement to run all the binary builds in the same process is that the binary we built and tested against could have different behavior to the one built by goreleaser if factors like flags are not identical. Ideally, the binary we test against should be the binary we ship in the release.

The third issue is that the github release creation via goreleaser seems to not allow us to leverage GitHub's built-in automatic release notes combined with additional content and would make creating good, automated release notes more convoluted. By significantly improving the detail included in release notes we can deprecate the manual changelog additions which is an extra manual time sink on every release. This paves the way for zero-touch deployments and consistently reducing the lead time to delivering new features.

A forth issue is that goreleaser is an additional tool required to install globally on a local machine to be able to replicate any CI issues. 

1. Create a new "dist" makefile target which builds the provider binaries for all platforms and zips them along with readme and licence, then produces checksums. This matches the existing behavour of goreleaser.
2. Call GH publish directly via GitHub action instead of via goreleaser - pushing to S3 and creating a GitHub release.
3. Use schema-tools to summarise changes since last release and add to the release notes.
4. Auto-generate release notes from commits, excluding pulumi-bot automated updates.
5. Add linting of workflows to detect issues earlier.

Resolves https://github.com/pulumi/pulumi-azure-native/issues/2239
- Disable publishing Java prerelease SDKs because Maven has no lock files or ability to exclude prerelease versions if specifying a range.
- Keep S3 upload in place for prereleases because other prerelease SDKs will attempt to download the provider binary from S3.

We're not integrating arm2pulumi release into the new dist process as there's additional unknown work it does to generate ruby scripts and commit them to https://github.com/pulumi/homebrew-tap/blob/master/Formula/arm2pulumi.rb

The arm2pulumi release name has been adjusted to try and match what the provider release title will be so it's not set incorrectly when goreleaser for arm2pulumi writes to the existing release. This path can only be tested on a non-pre prelease unfortunately because we've never had a goreleaser config for arm2pulumi prerelease 😞 

Asides:
- Make the schema-full and metadata a file-target which generates both files - but therefore disable parallel jobs in CI to avoid duplication of work.
- Fix missing prebuild dependency for the provider test.
- Fix Go SDK push action expression pointing to missing variable.
- Introduce actionlint GitHub actions linting to give us early indication of likely errors.
- Fix all actionlist errors.

Example of new release notes: https://github.com/pulumi/pulumi-azure-native/releases/tag/untagged-10ca26f7247767f7a887

### CODE REVIEW

**Comment by squaremo** [comment]
It's worth recording the motivation for this change -- what's the problem with goreleaser?

**Comment by squaremo** [comment]
> Disable publishing Java prerelease SDKs because Maven has no lock files or ability to exclude prerelease versions if specifying a range.

Does this belong here, or is it its own change? (I'm just going on the description, I may have misunderstood)

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -80,26 +99,20 @@ jobs:
...
 
-      - name: GoReleaser release
-        if: "!inputs.prerelease"
-        uses: goreleaser/goreleaser-action@v4
+      - name: Create GH Release
+        uses: softprops/action-gh-release@v1
         with:
-          args: -p 3 release --rm-dist --timeout 60m0s
-          version: latest
+          tag_name: v${{ inputs.version }}
```
Comment (thomas11) [inline .github/workflows/publish.yml:110]: Will we not create a double `vv` here? Sorry if I missed a sanitizing step elsewhere.

**Comment by danielrbradley** [comment]
@squaremo 
> It's worth recording the motivation for this change -- what's the problem with goreleaser?

Done - prepended to the description.

> Does this belong here, or is it its own change?

This could have been a small isolated fix prior to this PR, but was discovered in the course of testing this new release process so has ended up bundled in here.

**Review by thomas11 — APPROVED** [review]
Looks great. Listing the architectures several times is a bit annoying but they should be fairly static.

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -68,6 +69,24 @@ jobs:
...
+        uses: jaxxstorm/action-install-gh-release@v1.9.0
+        with:
+          repo: pulumi/schema-tools
+
+      - name: Get Schema Change Summary
+        id: schema-summary
+        shell: bash
+        run: |
+          # Get latest stable release. Return only first column from result (tag).
+          LAST_VERSION=$(gh release list --repo pulumi/pulumi-azure-native --exclude-drafts --exclude-pre-releases --limit 1 | cut -f1)
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply .github/workflows/publish.yml:85]: Oh I like this - much neater! TIL

File: .github/workflows/publish.yml (hunk 1 above)
Comment (danielrbradley) [reply .github/workflows/publish.yml:110]: Hmm, need to re-check this ... I think it'd be worth trying to normalise this earlier in the chain to make this easier to reason about.

**Review by squaremo — APPROVED** [review]
The changes make sense to me, and I like not having things done in two different ways :star: I also like the accompanying tidies. I have left some comments for consideration; I don't think any are blockers.

File: Makefile [hunk 3]
```diff
@@ -19,9 +19,10 @@ endif
 
 JAVA_GEN        := pulumi-java-gen
 
-# Specify which platform we're building for locally - default to using uname to detect
+# Specify which platform we're building for locally - default to matching go's platform
 # Valid inputs: darwin-arm64 darwin-amd64 linux-arm64 linux-amd64 windows-amd64
-LOCAL_PLAT      ?= $(shell echo $$(uname -s | tr "[:upper:]" "[:lower:]")-$$(uname -m))
+# Grab 4th section, replace "/" with "-"
+LOCAL_PLAT      ?= $(shell go version | cut -d ' ' -f4 | sed 's/\//-/')
```
[the remark this answers is not in the document]
Comment (squaremo) [reply Makefile]: (Leftover comment, resolving)

File: .github/workflows/publish.yml (hunk 1 above)
Comment (danielrbradley) [reply .github/workflows/publish.yml:110]: I've just pushed https://github.com/pulumi/pulumi-azure-native/pull/2222/commits/1f55b8df9571b2f8d16308d782fce03bbfed68a6 which should normalise this when entering via a tag push.

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -335,9 +352,15 @@ jobs:
           # Find go.mod's, get parent dir name, concat with version, set as a tag
           find */go.mod -type f -exec sh -c "dirname {} | xargs basename | xargs printf "%s/$VERSION" | xargs git tag" \;
 
-      - name: Publish to GitHub repository
+      - name: Publish to prerelease branch
+        if: inputs.prerelease == true
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply .github/workflows/publish.yml:358]: Unfortunately the natural inverse is:

```yaml
if: !inputs.prerelease
```

which is not valid yaml (leading `!`) ... so I put them both in explicitly to mirror each other.

File: .github/workflows/publish.yml [hunk 5]
```diff
@@ -260,6 +273,10 @@ jobs:
           ${{github.workspace}}/sdk/java
 
       - name: Publish Java SDK
+        # Maven doesn't use lock files and version ranges can't exclude prereleases.
+        # Therefore we can't publish prereleases without risking breaking customers.
+        # All other ecosystems handle prereleases correctly, Maven is the exception.
+        if: inputs.prerelease == false
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply .github/workflows/publish.yml:281]: That's not valid yaml ... yay yaml! 😄

**Comment by thomas11** [comment]
@danielrbradley Should we change the "Looking good!" message from schema-tools? IMO it's a bit weird first thing in the release notes.

File: Makefile [hunk 6]
```diff
@@ -158,17 +170,63 @@ bin/pulumi-java-gen: .pulumi-java-gen.version bin/pulumictl
...
+bin/windows-amd64/arm2pulumi.exe: TARGET := windows-amd64
+bin/%/arm2pulumi bin/%/arm2pulumi.exe: bin/pulumictl .make/provider_mod_download provider/cmd/arm2pulumi/* .make/arm2pulumi_prebuild $(PROVIDER_PKG)
+	cd provider && go build -o $(WORKING_DIR)/$@ $(VERSION_FLAGS) $(PROJECT)/provider/cmd/arm2pulumi
 
 bin/$(CODEGEN): bin/pulumictl .make/provider_mod_download provider/cmd/$(CODEGEN)/* $(PROVIDER_PKG)
 	cd provider && go build -o $(WORKING_DIR)/bin/$(CODEGEN) $(VERSION_FLAGS) $(PROJECT)/provider/cmd/$(CODEGEN)
 
-bin/$(PROVIDER): bin/pulumictl .make/provider_mod_download provider/cmd/$(PROVIDER)/*.go .make/provider_prebuild $(PROVIDER_PKG)
+# Writes schema-full.json and metadata-compact.json to bin/
+bin/schema-full.json bin/metadata-compact.json: bin/$(CODEGEN) $(SPECS) .make/versions_v1 .make/versions_deprecated
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply Makefile:190]: Interesting - not come across grouped targets. Yes, I think that's technically more accurate. Annoyingly, it's still not smart enough to realise both are outputted by the same command and still runs the command twice in parallel if using `--jobs 2` 😒

**Comment by danielrbradley** [comment]
@thomas11 yeah, I want to improve the schema-tools output. I've made a start by trying to make it testable: https://github.com/pulumi/schema-tools/pull/14

File: Makefile [hunk 7]
```diff
@@ -158,17 +170,63 @@ bin/pulumi-java-gen: .pulumi-java-gen.version bin/pulumictl
...
+bin/linux-arm64/$(PROVIDER): TARGET := linux-arm64
+bin/darwin-amd64/$(PROVIDER): TARGET := darwin-amd64
+bin/darwin-arm64/$(PROVIDER): TARGET := darwin-arm64
+bin/windows-amd64/$(PROVIDER).exe: TARGET := windows-amd64
+bin/%/$(PROVIDER) bin/%/$(PROVIDER).exe: bin/pulumictl .make/provider_mod_download provider/cmd/$(PROVIDER)/*.go .make/provider_prebuild $(PROVIDER_PKG)
+	@# check the TARGET is set
+	test $(TARGET)
 	cd provider && \
-		go build -o $(WORKING_DIR)/bin/$(PROVIDER) $(VERSION_FLAGS) $(PROJECT)/provider/cmd/$(PROVIDER)
+		export GOOS=$$(echo "$(TARGET)" | cut -d "-" -f 1) && \
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply Makefile:210]: True, but will leave as-is for sake of trying to get this over the line.

File: Makefile [hunk 8]
```diff
@@ -158,17 +170,63 @@ bin/pulumi-java-gen: .pulumi-java-gen.version bin/pulumictl
...
 	bin/pulumictl download-binary -n pulumi-language-java -v $(shell cat .pulumi-java-gen.version) -r pulumi/pulumi-java
 
-bin/arm2pulumi: bin/pulumictl .make/provider_mod_download provider/cmd/arm2pulumi/* .make/arm2pulumi_prebuild $(PROVIDER_PKG)
-	cp bin/schema-full.json provider/cmd/arm2pulumi
-	cp bin/metadata-compact.json provider/cmd/arm2pulumi
-	cd provider && go build -o $(WORKING_DIR)/bin/arm2pulumi $(VERSION_FLAGS) $(PROJECT)/provider/cmd/arm2pulumi
+LOCAL_ARM2PULUMI_PATH := bin/$(GOOS)-$(GOARCH)/$(LOCAL_ARM2PULUMI_FILENAME)
+
+bin/$(LOCAL_ARM2PULUMI_FILENAME): $(LOCAL_ARM2PULUMI_PATH)
+	cp $(LOCAL_ARM2PULUMI_PATH) bin
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply Makefile:176]: Yes, think you're right. Good catch. Interestingly, I can't repro this locally any more so maybe apple's fixed this on the latest release. Still worth fixing though 👍

**Pull request merged**